### PR TITLE
[SPARK-35612][SQL] Support LZ4 compression in ORC data source

### PR DESCRIPTION
**Pull request by fornaix**
### What changes were proposed in this pull request?

This PR aims to support LZ4 compression in the ORC data source.

### Why are the changes needed?

Apache ORC supports LZ4 compression, but we cannot set LZ4 compression in the ORC data source

**BEFORE**

```scala
scala> spark.range(10).write.option("compression", "lz4").orc("/tmp/lz4")
java.lang.IllegalArgumentException: Codec [lz4] is not available. Available codecs are uncompressed, lzo, snappy, zlib, none, zstd.
```

**AFTER**

```scala
scala> spark.range(10).write.option("compression", "lz4").orc("/tmp/lz4")
```
```bash
$ orc-tools meta /tmp/lz4                                                               
Processing data file file:/tmp/lz4/part-00000-6a244eee-b092-4c79-a977-fb8a69dde2eb-c000.lz4.orc [length: 222]
Structure for file:/tmp/lz4/part-00000-6a244eee-b092-4c79-a977-fb8a69dde2eb-c000.lz4.orc
File Version: 0.12 with ORC_517
Rows: 10
Compression: LZ4
Compression size: 262144
Type: struct<id:bigint>

Stripe Statistics:
  Stripe 1:
    Column 0: count: 10 hasNull: false
    Column 1: count: 10 hasNull: false bytesOnDisk: 7 min: 0 max: 9 sum: 45

File Statistics:
  Column 0: count: 10 hasNull: false
  Column 1: count: 10 hasNull: false bytesOnDisk: 7 min: 0 max: 9 sum: 45

Stripes:
  Stripe: offset: 3 data: 7 rows: 10 tail: 35 index: 35
    Stream: column 0 section ROW_INDEX start: 3 length 11
    Stream: column 1 section ROW_INDEX start: 14 length 24
    Stream: column 1 section DATA start: 38 length 7
    Encoding column 0: DIRECT
    Encoding column 1: DIRECT_V2

File length: 222 bytes
Padding length: 0 bytes
Padding ratio: 0%

User Metadata:
  org.apache.spark.version=3.2.0
```

### Does this PR introduce _any_ user-facing change?

Yes.

### How was this patch tested?

Pass the newly added test case.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcOptions.scala [hunk 1]
```diff
@@ -78,7 +78,8 @@ object OrcOptions {
     "snappy" -> "SNAPPY",
     "zlib" -> "ZLIB",
     "lzo" -> "LZO",
-    "zstd" -> "ZSTD")
+    "zstd" -> "ZSTD",
+    "lz4" -> "LZ4")
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcOptions.scala:82]: If you don't mind, please add this before `zstd` please?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala [hunk 2]
```diff
@@ -46,7 +46,8 @@ object OrcUtils extends Logging {
     "SNAPPY" -> ".snappy",
     "ZLIB" -> ".zlib",
     "ZSTD" -> ".zstd",
-    "LZO" -> ".lzo")
+    "LZO" -> ".lzo",
+    "LZ4" -> ".lz4")
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala:50]: Could you move this to before `LZO` and after `ZSTD`?

**Review by viirya — COMMENTED** [review]
What will happen if set `spark.sql.orc.impl` to `hive`? Does this lz4 option work?

**Comment by fornaix** [comment]
Thank you for review, @dongjoon-hyun 
It has been updated.

**Comment by fornaix** [comment]
> What will happen if set `spark.sql.orc.impl` to `hive`? Does this lz4 option work?

Sure, it does work.
But the lz4 codec name is missing in file names. I will fix it.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala [hunk 3]
```diff
@@ -314,6 +314,7 @@ private[orc] object OrcFileFormat extends HiveInspectors with Logging {
     "NONE" -> "",
     "SNAPPY" -> ".snappy",
     "ZLIB" -> ".zlib",
+    "LZ4" -> ".lz4",
```
Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala:317]: Can you also add test for "hive" case?

Comment (dongjoon-hyun) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala:317]: +1 for @viirya 's suggestion.

Comment (dongjoon-hyun) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala:317]: Please add it to `HiveOrcSourceSuite.scala`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala [hunk 4]
```diff
@@ -649,4 +649,14 @@ class OrcSourceSuite extends OrcSuite with SharedSparkSession {
       checkAnswer(spark.sql("SELECT _col0, _col2.c1 FROM t1"), Seq(Row(1, "a")))
     }
   }
+
+  test("SPARK-35612: Support LZ4 compression in ORC data source") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala:653]: Could you move this from `OrcSourceSuite` to `OrcSuite`? Then, `HiveOrcSourceSuite` will test it automatically.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala (hunk 3 above)
Comment (fornaix) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala:317]: +1, I will add a test case

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala (hunk 4 above)
Comment (fornaix) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala:653]: It's helpful, thanks

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @fornaix and @viirya .
Merged to master.

**Comment by dongjoon-hyun** [comment]
Thank you for your first contribution, @fornaix .
I added you to the Apache Spark contributor group and assigned SPARK-35612 to you.
Welcome to the Apache Spark community!

**Pull request closed without merging**